### ISSUE TITLE
Orleans 1.5.2 bug - GrainStateStorageBridge loses exception callstack

### ISSUE THREAD

**Issue by Eldar1205** [dotnet/orleans#4662]
Hi,

I've looked into code of `GrainStateStorageBridge` and in the method `WriteStateAsync` found that if an `OrleansException` is caught is won't be wrapped in an `OrleansException` and will later thrown as is in a way that loses the original callstack, which is a bug:

```
public async Task WriteStateAsync()
    {
      Stopwatch sw = Stopwatch.StartNew();
      GrainReference grainRef = this.baseGrain.GrainReference;
      Exception exception;
      try
      {
        await this.store.WriteStateAsync(this.grainTypeName, grainRef, this.statefulGrain.GrainState);
        StorageStatisticsGroup.OnStorageWrite(this.store, this.grainTypeName, grainRef, sw.Elapsed);
        exception = (Exception) null;
      }
      catch (Exception ex)
      {
        exception = ex;
      }
      if (exception != null)
      {
        StorageStatisticsGroup.OnStorageWriteError(this.store, this.grainTypeName, grainRef);
        string message = this.MakeErrorMsg("WriteState", exception);
        this.store.Log.Error(102203, message, exception);
        if (!(exception is OrleansException))
          exception = (Exception) new OrleansException(message, exception);
      }
      sw.Stop();
      if (exception != null)
        throw exception;
    }
```

**Comment by xiazen** [comment]
I'm a little lost. OrleansException will lose the original call stack if it didnt wrap a innerException, which is supposed to be the original exception. Is this correct? If yes, then this has nothing to do with GrainStateStorageBridge, but with how the original OrleansException is generated. Right ? 

Please correct me if I understand something terribly wrong.

**Comment by sergeybykov** [comment]
Is the issue here that if the caught exception is an `OrleansException` instead of doing `throw` from within the `catch` block we do `throw exception` from outside of it, and hence losing the call stack?

**Comment by Eldar1205** [comment]
Exactly

**Comment by xiazen** [comment]
ohhhhhh hence losing the original call stack. sorry I was being silly. If we move the 
```

 if (exception != null)
      {
        StorageStatisticsGroup.OnStorageWriteError(this.store, this.grainTypeName, grainRef);
        string message = this.MakeErrorMsg("WriteState", exception);
        this.store.Log.Error(102203, message, exception);
        if (!(exception is OrleansException))
          exception = (Exception) new OrleansException(message, exception);
      }
      sw.Stop();
      if (exception != null)
        throw exception;

```
back to the catch block, then we should be fine right?

**Comment by Eldar1205** [comment]
I think the stopwatch is the reason the code is outside the catch block,
can be refactored to a finally clause

**Comment by ReubenBond** [comment]
This is fixed in 2.0, does this warrant a 1.5.x patch release?
